File: src/pkg_py_example/pkg_py_example/simple_parameter.py
import rclpy  # 导入rclpy模块，这是ROS2的Python客户端库
from rclpy.node import Node  # 从rclpy.node模块中导入Node类，用于创建节点
from rcl_interfaces.msg import SetParametersResult  # 从rcl_interfaces消息中导入SetParametersResult消息类型
from rclpy.parameter import Parameter  # 从rclpy.parameter模块中导入Parameter类

# 定义一个名为SimpleParameter的类，继承自Node类
class SimpleParameter(Node):
    
    def __init__(self):
        super().__init__("simple_parameter")  # 调用父类的构造函数，创建节点，节点名称为simple_parameter
        self.declare_parameter("simple_int_param", 28)  # 声明一个整数类型的参数，默认值为28
        self.declare_parameter("simple_string_param", "Robot")  # 声明一个字符串类型的参数，默认值为"Robot"

        # 添加一个设置参数的回调函数，当参数发生变化时调用
        self.add_on_set_parameters_callback(self.paramChangeCallback)

    # 定义参数变化的回调函数
    def paramChangeCallback(self, params):
        result = SetParametersResult()  # 创建一个SetParametersResult对象

        # 遍历所有传递进来的参数
        for param in params:
            # 如果参数名称为simple_int_param并且类型为整数
            if param.name == "simple_int_param" and param.type_ == Parameter.Type.INTEGER:
                self.get_logger().info("Param simple_int_param changed! New value is %d" % param.value)  # 打印新值
                result.successful = True  # 设置结果成功

            # 如果参数名称为simple_string_param并且类型为字符串
            if param.name == "simple_string_param" and param.type_ == Parameter.Type.STRING:
                self.get_logger().info("Param simple_string_param changed! New value is %s" % param.value)  # 打印新值
                result.successful = True  # 设置结果成功

        # 返回结果对象
        return result

# 定义主函数
def main():
    rclpy.init()  # 初始化rclpy
    simple_parameter = SimpleParameter()  # 创建SimpleParameter对象
    rclpy.spin(simple_parameter)  # 保持节点运行，等待设置参数
    simple_parameter.destroy_node()  # 销毁节点
    rclpy.shutdown()  # 关闭rclpy

# 判断是否为主程序
if __name__ == "__main__":
    main()  # 如果是主程序，则运行主函数

File: src/pkg_cpp_example/src/simple_parameter.cpp
#include <rclcpp/rclcpp.hpp>                            // 导入rclcpp头文件，这是ROS2的C++客户端库
#include <rcl_interfaces/msg/set_parameters_result.hpp> // 导入rcl_interfaces消息中set_parameters_result消息类型

#include <vector> // 导入vector头文件，用于存储参数列表
#include <string> // 导入string头文件，用于字符串操作
#include <memory> // 导入memory头文件，用于智能指针

using std::placeholders::_1; // 使用placeholders命名空间中的_1占位符，用于回调函数

// 定义一个名为SimpleParameter的类，继承自rclcpp::Node
class SimpleParameter : public rclcpp::Node
{
public:
    // 构造函数
    SimpleParameter() : Node("simple_parameter")
    {
        declare_parameter<int>("simple_int_param", 28);                   // 声明一个整数类型的参数，默认值为28
        declare_parameter<std::string>("simple_string_param", "Antonio"); // 声明一个字符串类型的参数，默认值为"Antonio"
        // 添加一个设置参数的回调函数，当参数发生变化时调用
        param_callback_handle_ = add_on_set_parameters_callback(std::bind(&SimpleParameter::paramChangeCallback, this, _1));
    }

private:
    // 声明一个设置参数回调处理的共享指针
    OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;

    // 定义设置参数变化的回调函数
    rcl_interfaces::msg::SetParametersResult paramChangeCallback(const std::vector<rclcpp::Parameter> &parameters)
    {
        // 声明结果消息变量
        rcl_interfaces::msg::SetParametersResult result;
        for (const auto &param : parameters) // 遍历传递进来的参数列表，检查每个参数的名称和类型
        {
            // 如果找到名为"simple_int_param"且类型为整数的参数
            if (param.get_name() == "simple_int_param" && param.get_type() == rclcpp::ParameterType::PARAMETER_INTEGER)
            {
                RCLCPP_INFO_STREAM(get_logger(), "Param simple_int_param changed! New value is " << param.as_int()); // 打印出新的整数值
                result.successful = true;                                                                            // 设置结果对象为成功
            }
            // 如果找到名为"simple_string_param"且类型为字符串的参数
            if (param.get_name() == "simple_string_param" && param.get_type() == rclcpp::ParameterType::PARAMETER_STRING)
            {
                RCLCPP_INFO_STREAM(get_logger(), "Param simple_string_param changed! New value is " << param.as_string()); // 打印出新的字符串值
                result.successful = true;                                                                                  // 设置结果对象为成功
            }
        }

        return result; // 返回结果
    }
};

// 定义主函数
int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);                        // 初始化rclcpp
    auto node = std::make_shared<SimpleParameter>(); // 创建SimpleParameter对象
    rclcpp::spin(node);                              // 保持节点运行，等待设置参数
    rclcpp::shutdown();                              // 关闭rclcpp
    return 0;                                        // 程序正常退出
}
